File: Backend/app/blueprint/methods/fixed_point.py
import pandas as pd


def fixed_point_method(g, x0, tol=1e-7, max_iter=1000):
    """
    Fixed-Point Iteration Method to find a solution to x = g(x).

    Parameters:
    g : function
        The function to apply in the fixed-point iteration (x = g(x)).
    x0 : float
        Initial guess for the fixed-point.
    tol : float, optional
        Tolerance for convergence. Default is 1e-7.
    max_iter : int, optional
        Maximum number of iterations. Default is 1000.

    Returns:
    x : float
        The approximate fixed point.
    iterations : int
        The number of iterations performed.
    converged : bool
        Whether the method converged.
    result_array : list of dict
        A list containing the iteration details.
    """
    result_array = []
    x = x0

    for i in range(max_iter):
        x_new = g(x)
        error = abs(x_new - x)

        result = {
            'i': i + 1,
            'x': x_new,
            'g_x': g(x_new),
            'error': error
        }
        result_array.append(result)

        if error < tol:
            return x_new, i + 1, True, result_array  # Converged

        x = x_new  # Update for the next iteration

    return x, max_iter, False, pd.DataFrame(result_array)  # Did not converge within max_iter


# Example usage:
if __name__ == "__main__":
    import math


    # Define the function g(x) for which we are solving x = g(x)
    def g(x):
        return math.cos(x)  # Example: g(x) = cos(x)


    # Initial guess
    x0 = 0.5

    # Call the fixed-point method
    solution, iterations, converged, results = fixed_point_method(g, x0)

    # Display the results
    if converged:
        print(f"Converged to solution: {solution} in {iterations} iterations.")
    else:
        print(f"Did not converge within {iterations} iterations.")

    # Print iteration details
    for res in results:
        print(f"Iteration {res['i']}: x = {res['x']}, g(x) = {res['g_x']}, error = {res['error']}")

File: rust_methods/fixed_point.rs
use std::f64;

fn fixed_point_method<F>(g: F, x0: f64, tol: f64, max_iter: usize) -> (f64, usize, bool, Vec<(usize, f64, f64, f64)>)
where
    F: Fn(f64) -> f64,
{
    let mut result_array = Vec::new();
    let mut x = x0;

    for i in 0..max_iter {
        let x_new = g(x);
        let error = (x_new - x).abs();

        result_array.push((i + 1, x_new, g(x_new), error));

        if error < tol {
            return (x_new, i + 1, true, result_array); // Converged
        }

        x = x_new; // Update for the next iteration
    }

    // Did not converge within max_iter
    (x, max_iter, false, result_array)
}

fn print_results(result_array: &Vec<(usize, f64, f64, f64)>) {
    println!("{:<10} {:<15} {:<15} {:<15}", "Iter", "x", "g(x)", "Error");

    for (i, x, g_x, error) in result_array {
        println!("{:<10} {:<15.8} {:<15.8} {:<15.8}", i, x, g_x, error);
    }
}

fn main() {
    // Example usage:
    let g = |x: f64| x.cos(); // Example function g(x) = cos(x)
    let x0 = 0.5; // Initial guess
    let tol = 1e-7; // Tolerance for convergence
    let max_iter = 1000; // Maximum number of iterations

    let (x, iterations, converged, result_array) = fixed_point_method(g, x0, tol, max_iter);

    println!("Fixed point: {}", x);
    println!("Iterations: {}", iterations);
    println!("Converged: {}", converged);

    // Print results in table-like format
    print_results(&result_array);
}
